### PR TITLE
fix tokenize(): potential bug of splitting pretrained tokens with newly added tokens

### PR DESCRIPTION
**Pull request by askerlee**
In the tokenizer base class, `split_on_token()` attempts to split input text by each of the added tokens. Because it uses `text.split(tok)`, it may accidentally split a token in the pretrained vocabulary at the middle.

For example a new token "ht" is added to the vocabulary. Then "light" will be split into `["lig", ""]`. But as "light" is a token in the pretrained vocabulary, it probably should be left intact to be processed by `self._tokenize()`.

Hence in this pull request, `text.split()` is replaced with `re.split()`, which will split only at word boundaries (`[^A-Za-z0-9_]` in regular expression). This behavior can be enabled by specifying a new `tokenize()` argument: `additional_tokens_as_full_words_only=True` (default: False). If it's specified in `tokenizer.encode(text, ...)`, it will still take effect, as this argument will be passed down to `tokenize()`.

On languages that have no or different word boundaries as above (such as Chinese or Japanese), this behavior may produce undesirable results, and the user can revert to the old `text.split()` by not specifying `additional_tokens_as_full_words_only` (it will take the default value `False`).

An explanation of the argument `additional_tokens_as_full_words_only` has been added to the docstring of `tokenize()`. A test function `test_add_partial_tokens_tokenizer()` has been added to `tokenization_bert_test.py`.

### CODE REVIEW

File: pytorch_transformers/tokenization_utils.py [hunk 1]
```diff
@@ -589,16 +590,30 @@ def add_special_tokens(self, special_tokens_dict):
 
         return added_tokens
 
-    def tokenize(self, text, **kwargs):
+    def tokenize(self, text, split_additional_on_word_boundaries=False, **kwargs):
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: Seems good to m. I think this argument could be renamed `additional_tokens_as_full_words_only`

Comment (askerlee) [reply pytorch_transformers/tokenization_utils.py]: Thank @thomwolf for suggesting a good name! I've renamed that variable 😃

**Comment by thomwolf** [comment]
Great, thanks @askerlee

**Comment by askerlee** [comment]
Oh so it won't be merged? Anyway it's just a small issue @thomwolf

**Comment by thomwolf** [comment]
Thanks for the heads up. I forgot to follow up on this one.
It's good to merge indeed (rebased on mater).
cc @LysandreJik

**Comment by thomwolf** [comment]
Hum actually this seems to break a number of tokenization tests.
Do you want to give it a look @askerlee?

**Comment by askerlee** [comment]
@thomwolf sure. will fix it ASAP.

**Comment by thomwolf** [comment]
cc @LysandreJik who is working on fixing #2096 and following #2101 which are both related to this PR.

**Comment by LysandreJik** [comment]
Hey @askerlee, thanks for your pull request. I'm currently working on it and adapting it to all models, I've updated the tests so that they fit the current master so don't worry about it.

**Comment by askerlee** [comment]
@LysandreJik thank you so much! Sorry have been through very busy days. Do I need to do anything now?

**Pull request closed without merging**